File: lib/caml_core/caml_callback.c
#include <stdlib.h>
#include <stdio.h>

#include "caml/alloc.h"
#include "caml/memory.h"
#include "caml/mlvalues.h"
#include "caml/callback.h"

#include "caml_callback.h"

#define Input_Number "Number"
#define Input_Quit "Quit"
#define Input_Info "Info"
#define Input_Copy "Copy"

static value get_input(const char* message){
    static const value* get_input_closure = NULL;
    if (!get_input_closure) get_input_closure = caml_named_value("get_input");
    value input = caml_callback2 (*get_input_closure, caml_copy_string(message), Val_unit);
    return input;
}

int read_int(const char* message, const char* error_message){
    static const value* read_int_safe_closure = NULL;
    if (!read_int_safe_closure) read_int_safe_closure = caml_named_value("read_int");
    value read = caml_callback3(*read_int_safe_closure, caml_copy_string(message), caml_copy_string(error_message), Val_unit);
    return Int_val(read);
}

user_input_t get_user_input(const char* message){
    value input = get_input(message);
    user_input_t user_input;
    if (Is_block(input) && Field(input, 0) == caml_hash_variant(Input_Number) ) {
        user_input.choice = NUMBER;
        user_input.value.number_choose = Int_val(Field(input, 1));
    } else if (input == caml_hash_variant(Input_Info)) {
        user_input.choice = INFO;
    } else if (input == caml_hash_variant(Input_Copy)) {
        user_input.choice = COPY;
    } else {
        user_input.choice = QUIT;
    }
    return user_input;
}

File: lib/caml_core/caml_callback.ml
let rec get_input message () = 
  print_string message;
  match read_line () with
  | s when s |> int_of_string_opt |> Option.is_some -> `Number (int_of_string s)
  | "i" | "info" -> `Info
  | "q" | "quit" -> `Quit
  | "c" | "copy" -> `Copy
  | _ -> Printf.printf "Wrong option\n"; get_input message ()
;;

let rec read_int_safe message error_message () = 
  (print_string message);
  match read_int_opt ()with
  | Some n -> n
  | None -> print_string error_message; read_int_safe message error_message()
;;

let read_line_message () = 
  Printf.printf "Output path\n> ";
  read_line ()

let () = Callback.register "get_input" get_input;;
let () = Callback.register "read_int" read_int_safe;;
let () = Callback.register "read_line" read_line_message;;